File: 0103-二叉树的锯齿形层次遍历/Solution.java
import java.util.List;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        LinkedList<TreeNode> queue = new LinkedList<>();
        boolean flag = false;
        queue.offer(root);
        while (!queue.isEmpty()) {
            int count = queue.size();
            List<Integer> level = new ArrayList<>();
            for (int i = 0; i < count; i++) {
                TreeNode curr = queue.poll();
                level.add(curr.val);
                if (curr.left != null) {
                    queue.offer(curr.left);
                }
                if (curr.right != null) {
                    queue.offer(curr.right);
                }
            }
            if(flag){
                for (int i = 0, j=level.size()-1; i < j; i++, j--) {
                    int temp = level.get(i);
                    level.set(i, level.get(j));
                    level.set(j, temp);
                }
            }
            flag = !flag;
            res.add(level);
        }
        return res;
    }
}

File: 0103-二叉树的锯齿形层次遍历/solution.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        ret = []
        if not root:
            return ret
        queue = [root]
        reverse = False
        while queue:
            level = []
            for _ in range(len(queue)):
                tmp = queue.pop(0)
                level.append(tmp.val)
                if tmp.left:
                    queue.append(tmp.left)
                if tmp.right:
                    queue.append(tmp.right)
            if reverse:
                level = level[::-1]
            reverse = not reverse
            ret.append(level)
        return ret